src/newgrf.cpp

```diff
@@ -5131,14 +5131,35 @@ static void SkipAct1(ByteReader &buf)
 	GrfMsg(3, "SkipAct1: Skipping {} sprites", _cur.skip_sprites);
 }
 
+using CachedCallback = std::pair<uint16_t, SpriteGroupID>;
+static std::vector<CachedCallback> _cached_callback_groups; ///< Sorted list of cached callback result spritegroups.
+
+static const SpriteGroup *GetCallbackResultGroup(uint16_t value)
+{
+	/* Old style callback results (only valid for version < 8) have the highest byte 0xFF so signify it is a callback result.
+	 * New style ones only have the highest bit set (allows 15-bit results, instead of just 8) */
+	if (_cur.grffile->grf_version < 8 && GB(value, 8, 8) == 0xFF) {
+		value &= ~0xFF00;
+	} else {
+		value &= ~0x8000;
+	}
+
+	/* Find position for value within the cached callback list. */
+	auto it = std::ranges::lower_bound(_cached_callback_groups, value, std::less{}, &CachedCallback::first);
```

Comment:
Not sure we've decided we want ranges yet :p

Reply:
Oops, I accidentally already sneaked some in.

Why not, if it builds and makes things simpler to code?